### PR TITLE
feat: auto-scaling (part 1)

### PR DESCRIPTION
**Pull request by whynowy**
Signed-off-by: Derek Wang <whynowy@gmail.com>

This is the first part of the code changes for auto-scaling feature.

1. An `Ratable` interface is introduced in this PR, which will be implemented in the ISB reader and writer to measure the processing rate. For ISB reader, the `Rate()` returns the acknowledge rate, in ISB writer, it returns the writing rate.

2. An `Pending()` function is also introduced into `Reader` interface, which is intended for the ISB Reader and sources to implement.

### CODE REVIEW

File: pkg/shared/queue/overflow_queue.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+		maxSize:  size,
+		lock:     new(sync.RWMutex),
+	}
+}
+
+// Append adds an element to the queue
+func (q *OverflowQueue[T]) Append(value T) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+	if len(q.elements) >= q.maxSize {
```
Comment (vigith) [inline pkg/shared/queue/overflow_queue.go:24]: NIT: likely path is going to run this branch all the time.

File: pkg/isb/jetstream/writer.go [hunk 2]
```diff
@@ -134,6 +146,29 @@ func (jw *jetStreamWriter) Close() error {
 	return nil
 }
 
+// Rate returns the writting rate (tps)
+func (jw *jetStreamWriter) Rate(_ context.Context) (float64, error) {
+	if jw.opts.sequenceCheck {
```
Comment (vigith) [inline pkg/isb/jetstream/writer.go:151]: let's remove this.

File: pkg/isb/jetstream/writer.go [hunk 3]
```diff
@@ -134,6 +146,29 @@ func (jw *jetStreamWriter) Close() error {
...
+		return isb.RateNotAvailable, nil
+	}
+	timestampedSeqs := jw.writtenInfo.ReversedItems() // ReversedItems makes sure the latest is in the front
```
Comment (vigith) [inline pkg/isb/jetstream/writer.go:154]: this is O(N)

File: pkg/isb/jetstream/reader.go [hunk 4]
```diff
@@ -100,6 +109,67 @@ func (jr *jetStreamReader) Close() error {
...
+	c, err := jr.js.ConsumerInfo(jr.stream, jr.stream)
+	if err != nil {
+		return isb.PendingNotAvailable, fmt.Errorf("failed to get consumer info, %w", err)
+	}
+	return int64(c.NumPending), nil
+}
+
+// Rate returns the ack rate (tps)
+func (jr *jetStreamReader) Rate(_ context.Context) (float64, error) {
+	if !jr.opts.ackInfoCheck {
```
Comment (vigith) [inline pkg/isb/jetstream/reader.go:152]: please comment this is for sink. perhaps use `useAckInfoAsSinkRate` or something like this?

File: pkg/isb/jetstream/reader.go [hunk 5]
```diff
@@ -100,6 +109,67 @@ func (jr *jetStreamReader) Close() error {
...
+		return isb.RateNotAvailable, nil
+	}
+	timestampedSeqs := jr.ackedInfo.ReversedItems() // ReversedItems makes sure the latest is in the front
```
Comment (vigith) [inline pkg/isb/jetstream/reader.go:155]: O(N)

**Review by vigith — APPROVED** [review]
few comments

File: pkg/isb/jetstream/reader.go (hunk 5 above)
Comment (whynowy) [reply pkg/isb/jetstream/reader.go]: Changed to use Items().

File: pkg/isb/jetstream/reader.go (hunk 4 above)
Comment (whynowy) [reply pkg/isb/jetstream/reader.go]: done.

File: pkg/shared/queue/overflow_queue.go (hunk 1 above)
Comment (whynowy) [reply pkg/shared/queue/overflow_queue.go:24]: Will evaluate the performance of using container.List later.

**Pull request merged**